COLLABORATOR_GUIDE.md

```diff
@@ -73,6 +73,77 @@ All pull requests that modify executable code should be subjected to
 continuous integration tests on the
 [project CI server](https://ci.nodejs.org/).
 
+## Breaking Changes
+
+Backwards-incompatible changes may land on the master branch at any time after
+sufficient review by collaborators and approval of at least two CTC members.
+
+Examples of breaking changes include, but are not necessarily limited to,
+removal or redefinition of existing API arguments, changing return values,
+adding or removing errors, altering the timing, and changing the non-internal
+side effects of using a particular API.
+
+With a few notable exceptions outlined below, when backwards incompatible
+changes to an API are necessary, the existing API *must* be Deprecated *first*
+and the new API either introduced in parallel or added after the next major
+Node.js version following the Deprecation as a replacement for the Deprecated
+API. In other words, existing API's *must not* change without a Deprecation.
+
+Exception to this rule is given in the following cases:
+
+* Adding or removing errors thrown or reported by an API;
+* Changing error messages;
+* Altering the timing and non-internal side effects of the API.
+
+Such changes *must* be handled as semver-major changes but MAY be landed
+without a [Deprecation cycle](#deprecation-cycle).
+
+Purely additive changes (e.g. adding new events to EventEmitter 
+implementations, or adding new arguments to a method in a way that allows
+existing code to continue working without modification) are handled as
+semver-minor changes.
+
+## Deprecations
+
+Deprecation refers to the identification of API's that should no longer be used
+and that may be removed in a future major release of Node.js.
+
+Node.js uses three fundamental Deprecation levels:
+
+* Soft-Deprecation (or Docs-only deprecation) refers to elements of the Node.js
+  public API that are being staged for deprecation in a future Node.js major
+  release. An explicit notice indicating the deprecated status is added to the
+  API documentation *but no functional changes are implemented in the code*.
+  There will be no runtime deprecation warning.
+
+* Hard-Deprecation refers to the use of warnings being reported at
+  runtime when a deprecated API is used. By default, Node.js will print a
+  warning to stderr the first time a deprecated API is used. A command-line
+  switch can be used to escalate such warnings into runtime errors that will
+  cause the Node.js process to exit. As with Soft-Deprecation, the documentation
+  for the API must be updated to clearly indicate the deprecated status.
+
+* End-of-life refers to API's that have gone through Hard-Deprecation and are
+  ready to be removed from Node.js entirely.
+
+All Deprecations *must* be handled as breaking (semver-major) changes. This
+means that while a PR adding a Deprecation can be landed in master at any time
+with appropriate review and approval, such changes are only permitted to be
+delivered in new major releases.
```

Comment:
Agreed but it's going to be difficult to succinctly capture the nuance
there... hmm.. will give the wording around some more thought.

On Mon, Aug 8, 2016 at 9:29 AM, Jeremiah Senkpiel <notifications@github.com>
wrote:

> In COLLABORATOR_GUIDE.md
> <https://github.com/nodejs/node/pull/7964#discussion_r73907390>:
>
> > +  There will be no runtime deprecation warning.
> > +
> > +* Hard-Deprecation refers to the use of warnings being reported at
> > +  runtime when a deprecated API is used. By default, Node.js will print a
> > +  warning to stderr the first time a deprecated API is used. A command-line
> > +  switch can be used to escalate such warnings into runtime errors that will
> > +  cause the Node.js process to exit. As with Soft-Deprecation, the documentation
> > +  for the API must be updated to clearly indicate the deprecated status.
> > +
> > +* End-of-life refers to API's that have gone through Hard-Deprecation and are
> > +  ready to be removed from Node.js entirely.
> > +
> > +All Deprecations *must* be handled as breaking (semver-major) changes. This
> > +means that while a PR adding a Deprecation can be landed in master at any time
> > +with appropriate review and approval, such changes are only permitted to be
> > +delivered in new major releases.
>
>
>    - native realpath is accepted as a semver-major, we know it changes
>    internals but do not find any problems
>    - it releases in a semver-major (v6)
>    - afterwards it is discovered that it has side-effects which break too
>    much and could have been prevented
>    - a revert is made but we still plan to do this in some form in the
>    future
>
> Intention was already made in this major, unbreaking something but then
> printing a deprecation for it isn't breaking at this point, it's still
> unbreaking. We should be able to do that imo.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/7964/files/80fba8b5f1cb746626bc78939a9782be2cb81123#r73907390>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAa2eSEkK4QtcaDPUAzPzdJ0DWUJW0fAks5qd1ljgaJpZM4Jb68Q>
> .
>